src/chessboard/hooks/usePromotion.ts

```diff
@@ -0,0 +1,162 @@
+import { useState, useEffect, useCallback } from "react";
+import { Square, PromotionOption, Piece, Promotion } from "../types";
+
+export type Move = {
+  from: Square;
+  to: Square;
+  piece: Piece;
+  promotion?: PromotionOption;
+};
+
+type HandleMoveWithPossiblePromotion = (move: Move) => {
+  status: "success" | "illegal move" | "need promotion";
+};
+
+type PromotionState = {
+  isDialogOpen: boolean;
+  fromSquare: Square;
+  targetSquare: Square;
+  newPiece: PromotionOption;
+  piece: Piece;
+};
+
+const possiblePromotionFilesFromFile = new Map<string, Array<string>>([
+  ["a", ["a", "b"]],
+  ["b", ["a", "b", "c"]],
+  ["c", ["b", "c", "d"]],
+  ["d", ["c", "d", "e"]],
+  ["e", ["d", "e", "f"]],
+  ["f", ["e", "f", "g"]],
+  ["g", ["f", "g", "h"]],
+  ["h", ["g", "h"]],
+]);
+
+const getValidPawnMovesDefault = (square: Square): Array<Square> => {
+  const [squareFile, squareLine] = square;
+  const possibleFiles = possiblePromotionFilesFromFile.get(squareFile);
+  const possibleLine = Number(squareLine) === 7 ? 8 : 1;
+  if (!possibleFiles) return [];
+
+  return possibleFiles.map((file: string) => (file + possibleLine) as Square);
+};
+
+export const usePromotion = ({
+  onMakeMove,
+  getValidPawnMoves = getValidPawnMovesDefault,
+  autoPromoteToQueen = false,
+}: {
+  onMakeMove: (move: Move) => boolean;
+  getValidPawnMoves?: (square: Square) => Array<Square>;
+  autoPromoteToQueen?: boolean;
```

Comment:
I'm thinking also to add custom callback functions `onPromotionDialogOpened` and `onPromotionDialogClosed`, for user own side effects, but don't sure how useful they could be

Reply:
for now, leave them out as to not keep adding size to this PR, they can be considered in the future if asked for